common/producer/BaseProducer.java

```diff
@@ -20,42 +20,53 @@
 
 import grakn.core.common.iterator.ResourceIterator;
 
+import javax.annotation.concurrent.ThreadSafe;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 import static grakn.core.common.concurrent.ExecutorService.forkJoinPool;
 
+@ThreadSafe
 public class BaseProducer<T> implements Producer<T> {
 
     private final ResourceIterator<T> iterator;
+    private final AtomicBoolean isDone;
     private CompletableFuture<Void> future;
 
     BaseProducer(ResourceIterator<T> iterator) {
         this.iterator = iterator;
+        this.isDone = new AtomicBoolean(false);
         this.future = CompletableFuture.completedFuture(null);
     }
 
     @Override
-    public synchronized void produce(Queue<T> queue, int count) {
-        future = future.thenRunAsync(() -> produceAsync(queue, count), forkJoinPool());
+    public synchronized void produce(Queue<T> queue, int request) {
+        if (isDone.get()) return;
```

Comment:
does it need to be atomic given that `produce` is `synchronized`?

Reply:
`done(queue)` and `done(queue, error)` are not synchronised